File: swift-app/Sources/MacCleanerApp/NotificationService.swift
import UserNotifications
import Foundation

// Smart scheduling: auto-scan every N hours, notify if junk found > threshold
@MainActor
final class NotificationService {
    static let shared = NotificationService()

    private let center = UNUserNotificationCenter.current()

    // MARK: Permission

    func requestPermission() {
        center.requestAuthorization(options: [.alert, .sound, .badge]) { granted, error in
            if let error { print("[Notifications] permission error: \(error)") }
        }
    }

    // MARK: Scan result notification

    func notifyJunkFound(gb: Double, count: Int) {
        guard gb > 0.2 else { return }   // skip trivial amounts

        let content       = UNMutableNotificationContent()
        content.title     = "MacCleaner"
        content.body      = String(format: "%.1f GB of junk found (%d files). Tap to review.", gb, count)
        content.sound     = .default
        content.categoryIdentifier = "JUNK_FOUND"

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
        let request = UNNotificationRequest(identifier: "junk-found", content: content, trigger: trigger)
        center.add(request)
    }

    // MARK: Disk pressure notification

    func notifyLowDisk(freeGB: Double, usedPct: Double) {
        guard usedPct > 85 else { return }

        let content   = UNMutableNotificationContent()
        content.title = "⚠️ Low Disk Space"
        content.body  = String(format: "Only %.0f GB remaining. Open MacCleaner to free space.", freeGB)
        content.sound = .defaultCritical

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
        let request = UNNotificationRequest(identifier: "disk-pressure", content: content, trigger: trigger)
        center.add(request)
    }

    // MARK: Scheduled weekly scan

    func scheduleWeeklyScan() {
        center.removePendingNotificationRequests(withIdentifiers: ["weekly-scan"])

        var comps        = DateComponents()
        comps.weekday    = 2    // Monday
        comps.hour       = 10   // 10 AM

        let content      = UNMutableNotificationContent()
        content.title    = "MacCleaner Weekly Scan"
        content.body     = "Time for your weekly cleanup. Tap to scan now."
        content.sound    = .default

        let trigger = UNCalendarNotificationTrigger(dateMatching: comps, repeats: true)
        let request = UNNotificationRequest(identifier: "weekly-scan", content: content, trigger: trigger)
        center.add(request)
    }

    // MARK: Background auto-scan (runs every N hours via DispatchSource)

    private var scanTimer: DispatchSourceTimer?

    func startAutoScanTimer(intervalHours: Int = 2) {
        scanTimer?.cancel()
        let source = DispatchSource.makeTimerSource(queue: .global(qos: .background))
        source.schedule(deadline: .now() + .hours(intervalHours),
                        repeating: .hours(intervalHours))
        source.setEventHandler { [weak self] in
            Task { @MainActor in
                await self?.performBackgroundScan()
            }
        }
        source.resume()
        scanTimer = source
    }

    private func performBackgroundScan() async {
        let scanResp    = await ScanEngine.shared.scan()
        let storageResp = await ScanEngine.shared.storage()

        if let data = scanResp?.data, data.totalSizeGB > 0.2 {
            notifyJunkFound(gb: data.totalSizeGB, count: data.totalCount)
        }
        if let st = storageResp?.data {
            notifyLowDisk(freeGB: st.freeGB, usedPct: st.usedPct)
        }
    }
}

// MARK: - DispatchTimeInterval convenience

private extension DispatchTimeInterval {
    static func hours(_ h: Int) -> DispatchTimeInterval { .seconds(h * 3600) }
}

File: src/maccleaner/services/notification_service.py
"""macOS notification delivery via UserNotifications (pyobjc).

Falls back silently on non-macOS or missing permission — never crashes the app.
"""
from __future__ import annotations

import sys
import uuid


def _can_notify() -> bool:
    return sys.platform == "darwin"


def notify(title: str, body: str, identifier: str | None = None) -> None:
    """Send a macOS banner notification. No-op if permission not granted."""
    if not _can_notify():
        return
    try:
        import objc                                         # noqa: PLC0415
        from Foundation import NSObject                     # noqa: PLC0415
        from UserNotifications import (                     # noqa: PLC0415
            UNUserNotificationCenter,
            UNMutableNotificationContent,
            UNNotificationRequest,
            UNTimeIntervalNotificationTrigger,
        )

        center = UNUserNotificationCenter.currentNotificationCenter()

        content = UNMutableNotificationContent.alloc().init()
        content.setTitle_(title)
        content.setBody_(body)
        content.setSound_(None)

        trigger = UNTimeIntervalNotificationTrigger.triggerWithTimeInterval_repeats_(1, False)
        req_id  = identifier or str(uuid.uuid4())
        request = UNNotificationRequest.requestWithIdentifier_content_trigger_(
            req_id, content, trigger
        )
        center.addNotificationRequest_withCompletionHandler_(request, None)
    except Exception:
        pass


def request_permission() -> None:
    """Request notification permission on first launch (call once on startup)."""
    if not _can_notify():
        return
    try:
        from UserNotifications import (                     # noqa: PLC0415
            UNUserNotificationCenter,
            UNAuthorizationOptionAlert,
            UNAuthorizationOptionSound,
        )
        center = UNUserNotificationCenter.currentNotificationCenter()
        options = UNAuthorizationOptionAlert | UNAuthorizationOptionSound
        center.requestAuthorizationWithOptions_completionHandler_(options, None)
    except Exception:
        pass


def notify_scan_result(freed_gb: float, junk_gb: float) -> None:
    """Send a contextual scan-result notification."""
    if junk_gb < 0.1:
        return
    body = f"{junk_gb:.1f} GB of junk found — tap to review and clean."
    notify("MacCleaner", body, identifier="scan-result")


def notify_disk_pressure(free_pct: float) -> None:
    """Alert when disk is nearly full."""
    body = f"Only {free_pct:.0f}% disk space remaining. Tap to free space now."
    notify("MacCleaner — Low Disk Space ⚠️", body, identifier="disk-pressure")
